controller/controlplane/controller_reconciler_utils.go

```diff
@@ -120,3 +123,126 @@ func (r *Reconciler) ensureAdminMTLSCertificateSecret(
+	case operatorv2alpha1.WatchNamespacesTypeOwn:
+		return []string{cp.Namespace}, nil
+	case operatorv2alpha1.WatchNamespacesTypeList:
+		var nsList []string
+		for _, ns := range cp.Spec.WatchNamespaces.List {
+			if err := ensureWatchNamespaceGrantsForNamespace(ctx, r.Client, cp, ns); err != nil {
+				return nsList, err
+			}
+			nsList = append(nsList, ns)
+		}
+		// Add ControlPlane's own namespace as it will add it anyway because
+		// that's where the default "publish service" exists.
+		// We add it here as we do not require a ReferenceGrant for own namespace
+		// so there's no validation whether a grant exists.
+		nsList = append(nsList, cp.Namespace)
+
+		return nsList, nil
+	default:
+		return nil, fmt.Errorf("unexpected watchNamespaces.type: %q", cp.Spec.WatchNamespaces.Type)
+	}
+}
+
+// +kubebuilder:rbac:groups=gateway-operator.konghq.com,resources=watchnamespacegrants,verbs=list
+
+// ensureWatchNamespaceGrantsForNamespace ensures that a WatchNamespaceGrant exists for the
+// given namespace and ControlPlane.
+// It returns an error if a WatchNamespaceGrant is missing.
+func ensureWatchNamespaceGrantsForNamespace(
+	ctx context.Context,
+	cl client.Client,
+	cp *ControlPlane,
+	ns string,
+) error {
+	var grants operatorv1alpha1.WatchNamespaceGrantList
+	if err := cl.List(ctx, &grants, client.InNamespace(ns)); err != nil {
+		return fmt.Errorf("failed listing WatchNamespaceGrants in namespace %s: %w", ns, err)
+	}
+	for _, grant := range grants.Items {
+		if !watchNamespaceGrantContainsControlPlaneFrom(grant, cp) {
+			continue
+		}
+
+		return nil
+	}
+	return fmt.Errorf("WatchNamespaceGrant in Namespace %s to ControlPlane in Namespace %s not found", ns, cp.Namespace)
+}
+
+func watchNamespaceGrantContainsControlPlaneFrom(
+	grant operatorv1alpha1.WatchNamespaceGrant,
+	cp *ControlPlane,
+) bool {
+	for _, from := range grant.Spec.From {
+		if from.Group == operatorv2alpha1.SchemeGroupVersion.Group &&
+			from.Kind == "ControlPlane" &&
+			from.Namespace == cp.Namespace {
+			return true
+		}
+	}
+	return false
+}
+
+// validateWatchNamespaces validates that the operator's watch namespaces are compatible
+// with the ControlPlane's watch namespaces configuration.
+func validateWatchNamespaces(
+	cp *ControlPlane,
+	watchNamespaces []string,
+) error {
+	if cp.Spec.WatchNamespaces == nil {
+		return nil
+	}
+
+	// If ControlPlane is configured to watch all namespaces, operator should
+	// not be configured to watch any specific namespaces: it should have the watchNamespaces
+	// list empty, indicating that it will watch all namespaces.
+	switch cp.Spec.WatchNamespaces.Type {
+	case operatorv2alpha1.WatchNamespacesTypeAll:
+		if len(watchNamespaces) > 0 {
+			return fmt.Errorf(
+				"ControlPlane's watchNamespaces is set to 'All', but operator is only allowed on: %v",
+				watchNamespaces,
+			)
+		}
+
+	case operatorv2alpha1.WatchNamespacesTypeOwn:
+		if len(watchNamespaces) > 0 && !lo.Contains(watchNamespaces, cp.Namespace) {
```

Comment:
I think it is unlikely that a `ControlPlane`'s own namespace is not included in operator's watched namespaces but it is reconciled. It is OK to include the unusual case but I think we need to add some comments here and pay more attention to it.

Reply:
👍 Added some comments and extended the error message in 1b5721da57cf33581511431bc7b451df2deecddd

PTAL